notebooks/text_generator.livemd

```diff
@@ -0,0 +1,920 @@
+```elixir
+sequence_length = 100
+
+train_data =
+  normalized_book_text
+  |> Enum.map(&Map.fetch!(char_to_int, &1))
+  |> Enum.chunk_every(sequence_length, 1, :discard)
+  # We don't want the last chunk since we don't have a prediction for it.
+  |> Enum.slice(0..-2)
+  |> Nx.tensor()
+  |> Nx.divide(characters_count)
+  |> Nx.reshape({:auto, sequence_length, 1})
+```
+
+<!-- livebook:{"output":true} -->
+
+```
+#Nx.Tensor<
+  f32[155920][100][1]
+  EXLA.Backend<cuda:0, 0.1395165607.1902510098.152317>
+  [
+    [
+      [0.7586206793785095],
+      [0.3448275923728943],
+      [0.24137930572032928],
+      [0.03448275849223137],
+      [0.6206896305084229],
+      [0.6896551847457886],
+      [0.5862069129943848],
+      [0.41379308700561523],
+      [0.24137930572032928],
+      [0.17241379618644714],
+      [0.7586206793785095],
+      [0.03448275849223137],
+      [0.3103448152542114],
+      [0.7931034564971924],
+      [0.7586206793785095],
+      [0.24137930572032928],
+      [0.5517241358757019],
+      [0.13793103396892548],
+      [0.24137930572032928],
+      [0.6896551847457886],
+      [0.3103448152542114],
+      [0.03448275849223137],
+      [0.24137930572032928],
+      [0.13793103396892548],
+      [0.5862069129943848],
+      [0.5862069129943848],
+      [0.4482758641242981],
+      [0.03448275849223137],
+      [0.5862069129943848],
+      [0.27586206793785095],
+      [0.03448275849223137],
+      [0.10344827175140381],
+      [0.48275861144065857],
+      [0.37931033968925476],
+      [0.17241379618644714],
+      [0.24137930572032928],
+      [0.7241379022598267],
+      [0.03448275849223137],
+      [0.10344827175140381],
+      [0.20689654350280762],
+      [0.8275861740112305],
+      [0.24137930572032928],
+      [0.5517241358757019],
+      [0.7586206793785095],
+      [0.7931034564971924],
+      [0.6896551847457886],
+      [0.24137930572032928],
+      [0.7241379022598267],
+      [0.03448275849223137],
+      [0.37931033968925476],
+      ...
+    ],
+    ...
+  ]
+>
+```
+
+For our train results, We will do the same. Drop the first `sequence_length` characters and then convert them to the mapping. This time we need to do something called **one-hot encoding**.
+
+The reason we want to use one hot encoding is that in our model we don't want to only return a character as the output. We want it to return the possibility of each character for the output. This way we can decide if certain possibility is good or not or even we can decide between multiple possible outputs or even discard everything if the network is not confident enough.
+
+<!-- livebook:{"force_markdown":true} -->
+
+```elixir
+# How one hot encoding works example
+iex> one_hot_encode([
+iex>  0, 
+iex>  1, 
+iex>  2
+iex> ])
+[
+  [1, 0, 0],
+  [0, 1, 0],
+  [0, 0, 1]
+]
+```
+
+In Nx, you can achieve this encoding by using this snippet
+
+<!-- livebook:{"force_markdown":true} -->
+
+```elixir
+Nx.equal(tensor, Nx.tensor(Enum.to_list(0..last_feature)))
```

Comment:
```suggestion
Nx.equal(tensor, Nx.iota({1, last_feature}))
```

Maybe make it an anoonymous function, and show how it works instead of the iex snippet above?